### PR TITLE
fix(link-to): support angle-bracket invocation

### PR DESCRIPTION
**Pull request by buschtoens**
[RFC #0459](https://emberjs.github.io/rfcs/0459-angle-bracket-built-in-components.html) changed the attribute name from `targetRouteName` to `route`.

Here's the PR: https://github.com/emberjs/ember.js/pull/17772

This is currently breaking the Ember v3.10.0-beta series.

### CODE REVIEW

File: addon/components/link-to-external-component.js [hunk 1]
```diff
@@ -9,9 +9,11 @@ export default LinkComponent.extend({
     const owner = getOwner(this);
 
     if (owner.mountPoint) {
-      const targetRouteName = get(this, 'targetRouteName');
-      const externalRoute = owner._getExternalRoute(targetRouteName);
-      set(this, 'targetRouteName', externalRoute);
+      // https://emberjs.github.io/rfcs/0459-angle-bracket-built-in-components.html
```
Comment (villander) [inline addon/components/link-to-external-component.js]: can you add tests for this, please?

Comment (buschtoens) [reply addon/components/link-to-external-component.js]: We can't really specifically test for this, since this is dependent on the Ember version. However, `{{link-to}}` and `{{link-to-external}}` are tested extensively in the acceptance tests. I have verified, that in the `ember-beta` scenario without my fixes 25 tests are failing, while with my fixes applied only 1 unrelated test fails.

I plan to fix that test in a separate PR.

**Comment by buschtoens** [comment]
The following test is failing in L505:

https://github.com/ember-engines/ember-engines/blob/2ad3dfbe60409e884986749c89361d65a3a2d1c4/tests/acceptance/routeable-engine-demo-test.js#L490-L509

This is because of this line:

https://github.com/ember-engines/ember-engines/blob/2ad3dfbe60409e884986749c89361d65a3a2d1c4/tests/dummy/lib/ember-blog/addon/templates/post.hbs#L19

Links without an explicit `@route` link to the currently active route.

Relevant code from `<LinkTo>` in Ember core:

https://github.com/emberjs/ember.js/blob/2cc35738f2938e6b9735d0d641214b0721d53a4b/packages/%40ember/-internals/glimmer/lib/components/link-to.ts#L895-L900

https://github.com/emberjs/ember.js/blob/2cc35738f2938e6b9735d0d641214b0721d53a4b/packages/%40ember/-internals/glimmer/lib/components/link-to.ts#L490-L49

This then breaks the `href` computed property in L791:

https://github.com/emberjs/ember.js/blob/2cc35738f2938e6b9735d0d641214b0721d53a4b/packages/%40ember/-internals/glimmer/lib/components/link-to.ts#L749-L804

`routing.generateURL(route, models, query)` is called as:

```
routing.generateURL("blog.post.error", [], { lang: "Japanese" });
```

Which is missing the `error` param any error route expects, thus throwing this error:

```
Error: You must provide param `error` to `generate`.
```

AFAICT this should also throw in standalone Ember without any Engine interference. So this is either a regression, deprecated / removed or undefined behaviour.

**Comment by buschtoens** [comment]
Yup, I just confirmed that this is broken in Ember as well: https://codesandbox.io/s/p32ozvvz0x

File: addon/components/link-to-external-component.js (hunk 1 above)
Comment (villander) [reply addon/components/link-to-external-component.js]: I think that you can refactor this test with angle brackets - https://github.com/ember-engines/ember-engines/blob/master/tests/integration/component-with-link-to-external-test.js#L21-L26

since that your PR propose to do it, adds support to angle-bracket invocation

Comment (buschtoens) [reply addon/components/link-to-external-component.js]: Added angle bracket invocation tests for Ember `3.10.0-beta.1` and upwards.

**Comment by gokatz** [comment]
> Yup, I just confirmed that this is broken in Ember as well. I reported the bug here: emberjs/ember.js#17963

Just to confirm, is this still an issue in the 3.10 stable release? This is reproducible for me in 3.10

**Comment by buschtoens** [comment]
@stefanpenner People are repeatedly running into this bug and even a second PR #647 was submitted to fix this. While we are still waiting for https://github.com/emberjs/ember.js/issues/17963 to be fixed upstream, can we skip this edge-case test and cut a new release?

**Comment by scottkidder** [comment]
Any movement? This is blocking us and probably many others from 3.10

**Comment by buschtoens** [comment]
I'm using patch for [`patch-package`](https://github.com/ds300/patch-package/issues):

`./patches/ember-engines+0.7.1.patch`

```diff
diff --git a/node_modules/ember-engines/addon/components/link-to-component.js b/node_modules/ember-engines/addon/components/link-to-component.js
index 934c90c..f2a832d 100644
--- a/node_modules/ember-engines/addon/components/link-to-component.js
+++ b/node_modules/ember-engines/addon/components/link-to-component.js
@@ -8,7 +8,7 @@ export default LinkComponent.extend({
   didReceiveAttrs() {
     this._super(...arguments);

-    let owner = getOwner(this);
+    const owner = getOwner(this);

     assert(
       `You attempted to use {{link-to}} within a routeless engine, this is not supported. Use {{link-to-external}} to construct links within a routeless engine. See http://ember-engines.com/guide/linking-and-external-links for more info.`,
@@ -16,8 +16,11 @@ export default LinkComponent.extend({
     );

     if (owner.mountPoint) {
+      // https://emberjs.github.io/rfcs/0459-angle-bracket-built-in-components.html
+      const routeKey = 'targetRouteName' in this ? 'targetRouteName' : 'route';
+
       // Prepend engine mount point to targetRouteName
-      this._prefixProperty(owner.mountPoint, 'targetRouteName');
+      this._prefixProperty(owner.mountPoint, routeKey);

       // Prepend engine mount point to current-when if set
       if (get(this, 'current-when') !== null) {
@@ -26,8 +29,8 @@ export default LinkComponent.extend({
     }
   },

-  _prefixProperty(prefix, prop) {
-    let propValue = get(this, prop);
+  _prefixProperty(prefix, property) {
+    let propValue = get(this, property);

     // Sometimes `targetRouteName` will be a class
     if (typeOf(propValue) !== 'string') {
@@ -35,7 +38,7 @@ export default LinkComponent.extend({
     }

     let namespacedPropValue;
-    if (prop === 'current-when') {
+    if (property === 'current-when') {
       // `current-when` is a space-separated list of routes
       namespacedPropValue = propValue.split(' ');
       namespacedPropValue = namespacedPropValue.map(propValue =>
@@ -46,14 +49,14 @@ export default LinkComponent.extend({
       namespacedPropValue = this._namespacePropertyValue(prefix, propValue);
     }

-    set(this, prop, namespacedPropValue);
+    set(this, property, namespacedPropValue);
   },

   _namespacePropertyValue(prefix, propValue) {
     if (propValue === 'application') {
       return prefix;
-    } else {
-      return prefix + '.' + propValue;
     }
-  },
+      return prefix + '.' + propValue;
+
+  }
 });
diff --git a/node_modules/ember-engines/addon/components/link-to-external-component.js b/node_modules/ember-engines/addon/components/link-to-external-component.js
index bfb938a..a8756d7 100644
--- a/node_modules/ember-engines/addon/components/link-to-external-component.js
+++ b/node_modules/ember-engines/addon/components/link-to-external-component.js
@@ -9,9 +9,11 @@ export default LinkComponent.extend({
     const owner = getOwner(this);

     if (owner.mountPoint) {
-      const targetRouteName = get(this, 'targetRouteName');
-      const externalRoute = owner._getExternalRoute(targetRouteName);
-      set(this, 'targetRouteName', externalRoute);
+      // https://emberjs.github.io/rfcs/0459-angle-bracket-built-in-components.html
+      const routeKey = 'targetRouteName' in this ? 'targetRouteName' : 'route';
+      const routeName = get(this, routeKey);
+      const externalRoute = owner._getExternalRoute(routeName);
+      set(this, routeKey, externalRoute);
     }
-  },
+  }
 });
```

**Comment by raghavbk** [comment]
Any update? this is blocking us to upgrade ember 3.10

**Comment by buschtoens** [comment]
https://github.com/emberjs/ember.js/pull/17971 was just merged! 🎉 
The next `3.10.2` release should contain the fix.

**Comment by rwjblue** [comment]
v3.10.2 is being published by Ember's CI now, I'm going to go ahead and land this and release so folks are unblocked.

I'm sorry this has taken so long.

**Comment by villander** [comment]
@rwjblue we need update ember-cli here to get CI green?

**Comment by buschtoens** [comment]
@villander The issue should resolve itself with the next build, as `3.10.2` is now published.

**Comment by villander** [comment]
@buschtoens we just need rebuild this on CI?

**Comment by buschtoens** [comment]
AFAICT yes.

**Comment by rwjblue** [comment]
CI will be green after https://travis-ci.org/emberjs/ember.js/builds/547236303 is green and we restart ours.
